### ISSUE TITLE
parallel: please remove or fix

### ISSUE THREAD

**Issue by rkoe** [NixOS/nixpkgs#110584]
**Describe the bug**
GNU parallel (currently packaged as "parallel" and "parallel-full") is nagware, and violates both the GNU policy and the GPL.
(see also: https://bugs.launchpad.net/ubuntu/+source/parallel/+bug/1779764, https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=597050#75), and breaks scripts:

- It illegaly requires its users to
  - either cite an scientific article of the author if GNU parallel is used (directly or indirectly) for processing ones own data for a scientific article. This is:
    - illegal under GPL
    - illegal under GNU policies
    -  wrong from the scientific view / wrong according to scientific citing rules
       (Yes, academic tradition requires you to cite works you base your article on. But only using a software, which only speeds up the processing (and not even has any influence on the result), definitely does not fall in this category. It would be wrong to cite this software -- otherwise you would also have to cite your OS, bash, ls, your monitor and keyboard manufacturer etc.)
  - or pay 10000€ (which is also illegal under GPL and threatens and nags its users)
  - promise to cite (type 'will cite') to make the nagging message disappar.

- It breaks scripts which parse the output, since you have to run 'parallel --citation' for every user on every system to prevent that the scripts break.

In addition, there's a simpler, less buggy alternative "parallel" program in moreutils.

So, please:
- either remove parallel and parallel-full from the package-repository
- or at least move it to the "unfree"-part and fix the licence to non-free

**To Reproduce**
1. install GNU parallel: nix-env -iA nixos.parallel
2. run parallel: $parallel
   ```
   Academic tradition requires you to cite works you base your article on.
   If you use programs that use GNU Parallel to process data for an article in a
   scientific publication, please cite:

     Tange, O. (2020, August 22). GNU Parallel 20200822 ('Beirut').
     Zenodo. https://doi.org/10.5281/zenodo.3996295

   This helps funding further development; AND IT WON'T COST YOU A CENT.
   If you pay 10000 EUR you should feel free to use GNU Parallel without citing.

   More about funding GNU Parallel and the citation notice:
   https://www.gnu.org/software/parallel/parallel_design.html#Citation-notice

   To silence this citation notice: run 'parallel --citation' once.
   ```
   This is shown on every invocation, and breaks scripts.

3. uninstall GNU parallel again: nix-env -e parallel
4. install moreutils, if you need parallel: nix-env -iA nixos.moreutils

**Notify maintainers**
@pSub @vrthra @tomberek 

**Metadata**
 - system: `"x86_64-linux"`

**Comment by pstn** [comment]
Gnu parallel is licensed under [GPL 3.0](https://www.gnu.org/licenses/gpl-3.0.html). Could you please specify where exactly they are in violation of their license?

**Comment by SuperSandro2000** [comment]
This package is used in maintainers/scripts/check-maintainer-github-handles.sh and maintainers/scripts/update-luarocks-packages. If it should be removed you would need to find alternatives.

Also nixpkgs features unfree software which users need to opt in.

I would suggest you take this upstream and we close this for now.

**Comment by Profpatsch** [comment]
GNU parallel is annoying and I personally don’t think it’s good software (and the nag bar is the icing on the cake), but we shouldn’t dictate users whether they want to use this software or not.

I don’t think there is anything actionable here, so I’m inclined to close this issue if there is no counterargument from a maintainer.

**Comment by rkoe** [comment]
> Gnu parallel is licensed under [GPL 3.0](https://www.gnu.org/licenses/gpl-3.0.html). Could you please specify where exactly they are in violation of their license?

The author requires its users to cite his scientific article, or otherwise pay 10000€. This is not possible with GPL (and also violates the GNU policy, although it's a GNU project).
Yes, the author says that it's not a legal requirement (and so could be freely ignored), but this is a rather lame excuse, and the agressive wording and requiring to type 'will cite' tells something else.

> This package is used in maintainers/scripts/check-maintainer-github-handles.sh and maintainers/scripts/update-luarocks-

packages. If it should be removed you would need to find alternatives.
- `check-maintainer-github-handles.sh`: should be easy to convert to moreutils parallel
   (probably `| parallel -j5 checkUser` -> `| xargs parallel -j5 checkUser --`, but this has to be verified, of course)
- `update-luarocks-packages` would be more difficult

Question: How is the 'will cite' problem solved there?

> Also nixpkgs features unfree software which users need to opt in.

Yes, but GNU parallels is effectively unfree, since its license is effectively: GPLv3 + terms which are not allowed with GPLv3.
So, it (a) either has to go to unfree or (b) has to remove the additional terms ("cite or pay").

> I would suggest you take this upstream instead of moving the discussion making to downstream packaging.

Sorry, I guess there's no way to solve this upstream. The author knows this, but ignores it and uses lame excuses.

> > In addition, there's a simpler, less buggy alternative "parallel" program in moreutils.
> According to IRC this is not a drop in replacement.

Yes, it's not a drop in replacement, but a simpler variant (which according to the UNIX philosophy only does what it says), with less features but also less bugs.

**Comment by rkoe** [comment]
> we shouldn’t dictate users whether they want to use this software or not.

Agree.
So, I would suggest to move it to unfree, since that's what it is.

**Comment by Profpatsch** [comment]
> So, I would suggest to move it to unfree, since that's what it is.

Can you provide a list of what other distributions (especially free distributions according to the GNU project) list GNU parallel as? Then we will follow the consensus, or we can inquire if the licenses differ.

**Comment by 7c6f434c** [comment]
>> I would suggest you take this upstream instead of moving the discussion making to downstream packaging.
>Sorry, I guess there's no way to solve this upstream. The author knows this, but ignores it and uses lame excuses.

Well, if GNU project continues listing GNU parallel as a GNU project software package under GPL3+, it is not solely about the project author, and it does count as evidence that the citing requirements are not a part of legal license (and the package is free).

**Comment by gebner** [comment]
I don't interpret the annoying citation nagscreen as a requirement:
> If you use programs that use GNU Parallel [...], please cite:
> If you pay 10000 EUR you should feel free to use GNU Parallel without citing.

Neither of these two sentences says that you are not free to use GNU Parallel without citing.  (Note that it doesn't say "you must cite" or "only if you pay".)

The parallel source code also explicitly says that it is licensed under GPLv3+: https://git.savannah.gnu.org/cgit/parallel.git/tree/src/parallel#n6
We could legally patch out the nagscreen.

**Comment by grahamc** [comment]
>  We could legally patch out the nagscreen.

We could, but I don't think we should as a distribution take that position. They're asking for citations. If users are annoyed by that, `parallel` provides at least one way to silence it. If users are still annoyed they need to silence it, they could use something else.

**Comment by Profpatsch** [comment]
> If users are still annoyed they need to silence it, they could use something else or patch it themselves.

Since this is nix, we are technically able to provide a flag `disableNagMessage` that does it and people can set it with `parallel.override { disableNagMessage = true; }`.

I think that would be a good compromise. Please feel free to submit a PR with the patch and the flag @rkoe (or anybody else of course).

**Comment by rkoe** [comment]
I'm wondering why this was closed so fast, without being resolved.

TL;DR:
- I think NixOS should remove/reduce the nag-message by default, by one of:
   - patching (like several other distributions)
   - adding an empty "~/.parallel/will-cite" file by default -- but I think this is not possible with Nix
   - adding a wrapper which adds "--will-cite"

  In my opinion, a small patch would be best.
- Adding an optional, by default disabled flag "disableNagMessage" does not make any sense -- except if it is set to true by default. And then, a simple patch would be better.

-----

About other distributions:
- Debian: had an discussion about removing it completely or moving it to non-free, and then has patched out the citation notice to make GNU parallel GPL-compliant.
    https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=884793
    https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=905674
    https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=597050
    https://tracker.debian.org/news/1008286/accepted-parallel-20161222-11-source-into-unstable/

- Arch Linux: has patched out the citation notice and renamed it from GNU parallel to parallel:
  https://github.com/archlinux/svntogit-community/tree/packages/parallel/trunk
- OpenSUSE: has patched out the citation notice:
  https://build.opensuse.org/package/view_file/openSUSE:Factory/gnu_parallel/parallel-remove-nag-screen.patch?expand=0
- Ubuntu: has an open, confirmed bugreport
  https://bugs.launchpad.net/ubuntu/+source/parallel/+bug/1779764
- Fedora: probably uses upstream version
- Linux Mint: unknown; I have not found a "parallel"-package for Linux Mint.
- others: I have not checked


Details:
- Several other distributions have patched out the citation/payment notice. I think this is the way to go.
  If it is patched out completely, or only reduced (to only kindly ask for citation, similar to e.g. https://lists.gnu.org/archive/html/parallel/2013-11/msg00006.html), may be discussed.
  In either way, the request for paying 10000 EUR both is completely inapropriate and probably violates GPL,
  and should always be removed.
  (except if the author switches to a different licence, which allows such requirements)

- GPLv3 and required vs. optional:
  - Does the GPLv3 allow to politely ask for support/citation/donation? Yes.
  - Does the GPLv3 allow to require citation/payment? No.

  So the question is: Is the citation/payment required or purely optional?

  - From what I've read from the author of GNU parallel, Mr. Tange, he considers it as required (cite, pay or don't use the program) and *not* optional (cite or pay or ignore).
  - The problem is, that it's not allowed with GPLv3.
    So, Mr. Tange tries to sidestep the GPLv3, so he can use rules which are forbidden by the GPLv3.
    He often emphasizes, that "the wording was chosen carefully"; effectively with the goal to sidestep GPLv3, so that the citation/payment is not part of the license (which forbids it), but nevertheless is a requirement or at least understood as a requirement by the users.

    But: It's either optional or required -- having it required for the users but optional for legal reasons for the licence does not work. And the program should then also clearly state if it is optional or required.

    The current wording says that it's required: Please cite or pay 10000 EUR.
    (Previous manpage-versions even contained the sentence: "If you use '--will-cite' in scripts you are expected to pay the 10000 EUR, because you are making it harder to see the citation notice.")

**Comment by Profpatsch** [comment]
Well, that’s great, let’s reuse the patches from one of the other distros then. We are not violating the license of an open source project if we patch it (as long as we don’t touch the license) and there seems to be good precedence.

**Comment by tomberek** [comment]
### Citation Nag
There is some precedence in packages like Minecraft where the EULA agreement is a specific option that the Nix user must switch from the default to true. Something like that for the "--will-cite" wrapper is the equivalent. (as @Profpatsch mentioned)

## GPLv3, required/optional

From https://www.gnu.org/software/parallel/parallel_design.html
```
As the request for citation is not a legal requirement this is acceptable under GPLv3 and cleared with Richard M. Stallman himself. Thus it does not fall under this: https://www.gnu.org/licenses/gpl-faq.en.html#RequireCitation
```

This seems pretty definitive.

## Buggy and replacements
GNU Parallel is far better and has more features than any alternative I've come across (i've tried many of them, did a lot of ecosystem exploration). It's primary drawback is the high startup time per job; 10ms per job. It's fine to use moreutils-parallel for simple internal scripting uses in nixpkgs, but for day-to-day usage GNU Parallel is far superior.

## Conclusion
I've worked with the package as well as with upstream, would be willing to help coordinate/implement a solution.

**Comment by rkoe** [comment]
> ### Citation Nag
> 
> There is some precedence in packages like Minecraft where the EULA agreement is a specific option that the Nix user must switch from the default to true. Something like that for the "--will-cite" wrapper is the equivalent.

No, this is different. GNU parallel wants to use GPLv3 -- and GPLv3 does not allow further requirements or an additional EULA. And it's also not allowed to require the user to confirm additional, GPLv3-violating conditions.

If GNU parallel would use a different licence, which allows such additional requirements, then this bugreport would not exist.

> ## GPLv3, required/optional
> 
> From https://www.gnu.org/software/parallel/parallel_design.html
> 
> ```
> As the request for citation is not a legal requirement this is acceptable under GPLv3 and cleared with Richard M. Stallman himself. Thus it does not fall under this: https://www.gnu.org/licenses/gpl-faq.en.html#RequireCitation
> ```
So, I can add any additional requirements to GPL-programs, as long as I say "oh, by the way it's not a legal requirement -- but of course, it's required"?
Sorry, but if it's not a requirement, it should not try to tell the user that it's a requirement.

> ## Buggy and replacements
> 
> GNU Parallel is far better and has more features than any alternative I've come across

Yes, it has more features. But this also makes it very complex -- and it doesn't even get quoting right (!).

> ## Conclusion
> I've worked with the package as well as with upstream, would be willing to help coordinate/implement a solution.

Glad to hear. I would be happy with every solution, which does not try to fool the user.

**Comment by SuperSandro2000** [comment]
> Debian: had an discussion about removing it completely or moving it to non-free, and then has patched out the citation notice to make GNU parallel GPL-compliant.

If you would create a PR with that I would think this is okay.

**Comment by ole-tange** [comment]
It seems there is some confusion as to:

* whether the citation notice breaks scripts (It does not. People wrongly claim so but not a single time has that been shown to happen. https://git.savannah.gnu.org/cgit/parallel.git/tree/doc/citation-notice-faq.txt#n38)
* whether the notice is GPL3 compliant (It is according to RMS https://git.savannah.gnu.org/cgit/parallel.git/tree/doc/citation-notice-faq.txt#n27)
* whether you can patch out the notice and thus fork the program (You can, but then you cannot call the program GNU Parallel nor can use "GNU Parallel" in the description or documentation: https://git.savannah.gnu.org/cgit/parallel.git/tree/doc/citation-notice-faq.txt#n110).
* whether it is preferable to not be distributed at all (or distributed as non-free) over not having the notice (It is preferable not to be distributed at all https://git.savannah.gnu.org/cgit/parallel.git/tree/doc/citation-notice-faq.txt#n133)

Please read: https://git.savannah.gnu.org/cgit/parallel.git/tree/doc/citation-notice-faq.txt

Especially the last section is relevant.
